pkg/plugins/scorer/no_hit_lru.go

```diff
@@ -0,0 +1,263 @@
+
+	// defaultLRUSize is the maximum number of pods we'll consider in the cache
+	defaultLRUSize = 1024
+)
+
+// compile-time type assertions
+var _ framework.Scorer = &NoHitLRU{}
+var _ requestcontrol.PreRequest = &NoHitLRU{}
+
+// NoHitLRUParameters defines the parameters for the NoHitLRU scorer.
+type NoHitLRUParameters struct {
+	// PrefixPluginName defines the name of the prefix cache plugin to read state from.
+	// Defaults to "prefix-cache-scorer".
+	PrefixPluginName string `json:"prefixPluginName"`
+
+	// LRUSize defines the maximum number of pods to track in the LRU cache.
+	LRUSize int `json:"lruSize"`
+}
+
+// coldRequestState tracks whether a request triggered a KV cache hit
+// when the cache is missed, isCold is true.
+type coldRequestState struct {
+	isCold bool
+}
+
+// Clone implements the plugins.StateData interface
+func (c *coldRequestState) Clone() plugins.StateData {
+	return &coldRequestState{isCold: c.isCold}
+}
+
+// NoHitLRUFactory defines the factory function for the NoHitLRU
+func NoHitLRUFactory(name string, rawParameters json.RawMessage, handle plugins.Handle) (plugins.Plugin, error) {
+	parameters := NoHitLRUParameters{}
+	if rawParameters != nil {
+		if err := json.Unmarshal(rawParameters, &parameters); err != nil {
+			return nil, fmt.Errorf("failed to parse the parameters of the '%s' scorer - %w", NoHitLRUType, err)
+		}
+	}
+
+	if parameters.PrefixPluginName == "" {
+		parameters.PrefixPluginName = prefix.PrefixCachePluginType
+	}
+
+	// Note: We don't enforce that the prefix plugin exists here
+	// The scorer will gracefully handle missing prefix cache state as an optimization
+
+	return NewNoHitLRU(handle.Context(), &parameters).WithName(name), nil
+}
+
+// NewNoHitLRU creates a new NoHitLRU scorer
+func NewNoHitLRU(ctx context.Context, params *NoHitLRUParameters) *NoHitLRU {
+	prefixPluginName := prefix.PrefixCachePluginType
+	lruSize := defaultLRUSize
+
+	if params != nil {
+		if params.PrefixPluginName != "" {
+			prefixPluginName = params.PrefixPluginName
+		}
+		if params.LRUSize > 0 {
+			lruSize = params.LRUSize
+		}
+	}
+
+	lruCache, err := lru.New[string, struct{}](lruSize)
+	if err != nil {
+		log.FromContext(ctx).Error(err, fmt.Sprintf("failed to initialize NoHitLRU scorer: could not create LRU cache with size %d: %v", lruSize, err))
+		return nil
+	}
+
+	return &NoHitLRU{
+		typedName:        plugins.TypedName{Type: NoHitLRUType},
+		lruCache:         lruCache,
+		mutex:            &sync.RWMutex{},
+		prefixPluginName: prefixPluginName,
+		pluginState:      plugins.NewPluginState(ctx),
+	}
+}
+
+// NoHitLRU scorer that favors pods that were least recently used for cold requests.
+// This can help evenly distribute cache growth, since cold requests result in more
+// new KV blocks.
+type NoHitLRU struct {
+	typedName        plugins.TypedName
+	lruCache         *lru.Cache[string, struct{}] // pod name -> dummy value (we only care about order)
+	mutex            *sync.RWMutex
```

Comment:
the `lru.Cache` package is thread-safe, the `Add` and `Keys` calls are protected by an internal lock. The use of this mutex is therefore purely for snapshot semantics: If `Keys()` is called, the mutex makes sure no other request will progress before the scoring. Might be good to test the effect of this locking and profile the overhead as well.